### PR TITLE
test: add e2e tests for CTB restarts from file changes 

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- adds e2e test for creating a collection type and that the ctb tutorial modal appears
- adds util to watch for the server restart
- adds util to help ignore the ctb tutorial modal
- adds a very hacky test framework and util for resetting the filesystem state after changes like CTB edits, uploads, etc
- minor refactoring and docs

I'm not super happy with how hacky it is, but it seems stable enough from the time I've been working with it. The most important thing I would like to find a solution for is the `delay()` to wait for the server to realize files have changed and it begins its restart process. However, without access to the output of the server I can't think of anything better. Perhaps a custom winston logger could pipe it somewhere that we could watch that would be realtime?

### Why is it needed?

We were limited by being unable to test the CTB because of the server restarts, filesystem changes, etc.

### Roadmap

After this PR, I am planning the following changes:

- bring CLI tests from `v5/main` to `develop` (and then work from there on further changes)
- copy/paste and clean up first attempt at CTB tests from https://github.com/strapi/strapi/pull/18065
- refactor e2e+cli tests to share code, ideally everything except the test runner itself (jest vs playwright command)
- lots of other little cleanups, like improving the restart monitoring being added here

### How to test it?

Running the full test suite with multiple browsers (that is, re-using the same running instance of strapi across multiple test runs of the same suite, and re-using the test-apps between runs ie without using `--setup`) should work every time. And continuing to run them many times in a row, they should always succeed without flakiness.

### Related issue(s)/PR(s)

DX-1166

### CODE REVIEW

**Comment by alexandrebodin** [comment]
FYI there is also this historical PR with a bunch of good tests https://github.com/strapi/strapi/pull/18065

**Comment by innerdvations** [comment]
> FYI there is also this historical PR with a bunch of good tests #18065

I remember that one, maybe I can cannibalize the tests from there as soon as we have an acceptable way to run them. I want to see if I can eliminate the worst hacks in this PR, which should be possible if I manage to get even one-way direct communication between the playwright server process and the running tests (rather than using the admin API like we're doing now)

File: test/scripts/run-e2e-tests.js [hunk 1]
```diff
@@ -58,10 +58,10 @@ const setupTestEnvironment = async (generatedAppPath) => {
 yargs
   .parserConfiguration({
     /**
-     * This lets us pass any other arguments to playwright
-     * e.g. the name of a specific test or the project we want to run
+     * When unknown options is false, using -- to separate playwright args from test:e2e args works
+     * When it is true, the script gets confused about additional arguments, with or without using -- to separate commands
      */
-    'unknown-options-as-args': true,
+    'unknown-options-as-args': false,
```
Comment (innerdvations) [inline test/scripts/run-e2e-tests.js:64]: This and the corresponding update to the contributor doc should really be a separate PR, but without the ability to only run one file from one domain, it's extremely time consuming to test this, so I've just included it here.

Comment (joshuaellis) [reply test/scripts/run-e2e-tests.js:64]: > but without the ability to only run one file from one domain

You can do this `yarn test:e2e --domains=admin -- transfer`

Comment (innerdvations) [reply test/scripts/run-e2e-tests.js:64]: no, that's what I mean. This has to be flipped to false or else that doesn't work, at least for me. The previous comment+value appears to have been a lie

**Comment by innerdvations** [comment]
I'm going ahead and opening this PR as-is with the `delay()` giving strapi time to detect file changes and restart. While it could technically be flakey (it could take longer than 3s to being the restart process) I haven't seen it fail in about 5 runs on github so far, and never locally either with many more runs.

I'm very open to better ideas, but trying to watch the logs for the restart / started also feels overly complicated and potentially flakey as well.

File: tests/e2e/scripts/dts-import.js [hunk 2]
```diff
@@ -11,6 +12,59 @@ const {
...
+    }
+  }
+
+  // If we've exited the loop because of the timeout
+  console.error('Timeout reached, service did not become available in time.');
+}
+
+const gitUser = ['-c', 'user.name=Strapi CLI', '-c', 'user.email=test@strapi.io'];
+
+export const resetFiles = async () => {
```
Comment (joshuaellis) [inline tests/e2e/scripts/dts-import.js:48]: im not sure this should be in this file, this feels like it's own script? Or you could just change the name idm. Could do a few things really...

File: tests/e2e/tests/content-type-builder/collection-type/create-collection-type.spec.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    if (modal.isVisible()) {
+      await modal.click();
+      await expect(modal).not.toBeVisible();
+    }
+  });
+
+  // TODO: each test should have a beforeAll that does this, maybe combine all the setup into one util to simplify it
+  // to keep other suites that don't modify files from needing to reset files, clean up after ourselves at the end
+  test.afterAll(async () => {
+    await resetFiles();
```
Comment (joshuaellis) [inline tests/e2e/tests/content-type-builder/collection-type/create-collection-type.spec.ts:26]: why are you doing it afterAll but also in `beforeEach`? Maybe it should be in `afterEach` instead?

**Review by joshuaellis — COMMENTED** [review]
Overall looks great, well done for cracking this, super hard stuff. Just some minor comments to discuss

File: tests/e2e/scripts/dts-import.js (hunk 2 above)
Comment (innerdvations) [reply tests/e2e/scripts/dts-import.js:48]: oh, yeah, I didn't even pay attention to the filename, I was thinking it was a general utils, good catch 😆

File: tests/e2e/tests/content-type-builder/collection-type/create-collection-type.spec.ts (hunk 3 above)
Comment (innerdvations) [reply tests/e2e/tests/content-type-builder/collection-type/create-collection-type.spec.ts:26]: It's a tricky question here.

In this file, the correct thing would be only the `beforeEach` (since each test makes changes) and then other suites would ensure on their own it was clean with a beforeAll/beforeEach as needed.  But doing the filesystem reset is slow and only a couple suites make file changes (ctb and uploading), and I didn't want to update every other test file to do that.

So I put the "bad" code (trying to clean up for other tests in afterAll is bad practice because you're not guaranteed it will run) separate even though it means we run it one extra time, because ultimately I want to remove it altogether once we have a faster method to detect/reset the files that would be ok to add universally.

File: tests/e2e/scripts/dts-import.js (hunk 2 above)
Comment (innerdvations) [reply tests/e2e/scripts/dts-import.js:48]: I'll also move those three "scripts" to "utils" since that's what they are. It doesn't make sense to have both folders.

**Review by joshuaellis — APPROVED** [review]

**Comment by joshuaellis** [comment]
V nice well done

**Comment by innerdvations** [comment]
> V nice well done

not quite, one of the runs failed 😞

I think I have another solution though that I'll try to run through the CI several times and see if it still happens
